lib/rmg/WaterRoutes.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * WaterRoutes.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#include "WaterRoutes.h"
+#include "WaterProxy.h"
+#include "CMapGenerator.h"
+#include "RmgMap.h"
+#include "../mapping/CMap.h"
+#include "../mapping/CMapEditManager.h"
+#include "../mapObjects/CObjectClassesHandler.h"
+#include "RmgPath.h"
+#include "RmgObject.h"
+#include "ObjectManager.h"
+#include "Functions.h"
+#include "RoadPlacer.h"
+#include "TreasurePlacer.h"
+#include "TownPlacer.h"
+#include "ConnectionsPlacer.h"
+#include "TileInfo.h"
+#include "WaterAdopter.h"
+#include "RmgArea.h"
+
+void WaterRoutes::process()
+{
+	auto * wproxy = zone.getModificator<WaterProxy>();
+	if(!wproxy)
+		return;
+	
+	for(auto & z : map.getZones())
+	{
+		if(z.first != zone.getId())
+			result.push_back(wproxy->waterRoute(*z.second));
+	}
+	
+	//prohibit to place objects on sealed off lakes
+	for(auto & lake : wproxy->getLakes())
+	{
+		if((lake.area * zone.freePaths()).getTilesVector().size() == 1)
+		{
+			zone.freePaths().subtract(lake.area);
+			zone.areaPossible().subtract(lake.area);
+		}
+	}
+	
+	//prohibit to place objects on the borders
+	for(auto & t : zone.area().getBorder())
+	{
+		if(zone.areaPossible().contains(t))
+		{
+			std::vector<int3> landTiles;
+			map.foreachDirectNeighbour(t, [this, &landTiles, &t](const int3 & c)
+			{
+				if(map.isOnMap(c) && map.getZoneID(c) != zone.getId())
+				{
+					landTiles.push_back(c - t);
+				}
+			});
+			
+			if(landTiles.size() == 2)
+			{
+				int3 o = landTiles[0] + landTiles[1];
+				if(o.x * o.x * o.y * o.y == 1) 
```

Comment:
Hard to understand this, what does it mean? Why we check exactly 0 and 1?

Reply:
Is it possible that this expression will be 0 or 4?

Reply:
It checks for corner pattern. It could be zero, it water is 1-tile width - we shouldn't block placement in that case